gpii/node_modules/userListeners/src/listeners.js

```diff
@@ -0,0 +1,176 @@
+        stopListener: "fluid.notImplemented",
+        // Parse a Buffer containing an NDEF message, returning a string.
+        decodeNDEF: {
+            funcName: "gpii.userListeners.decodeNDEF",
+            args: [
+                "{arguments}.0" // NDEF buffer
+            ]
+        },
+        // The listener has failed, and should be restarted.
+        failed: {
+            funcName: "gpii.userListeners.failed",
+            args: [
+                "{that}",
+                "{arguments}.0" // The error.
+            ]
+        }
+    },
+    members: {
+        // Set to true to call proximityTriggered, otherwise login/logout.
+        proximity: false,
+        listenerName: "@expand:fluid.identity({that}.typeName)",
+        // Number of failures.
+        failCount: 0,
+        // Seconds (multiplied by failureCount) to wait before restarting.
+        failDelay: 10
+    },
+    listeners: {
+        // Start/stop this listener
+        "{userListeners}.events.onListenersStart":  "{that}.startListener",
+        "{userListeners}.events.onListenersStop": "{that}.stopListener",
+
+        "onTokenArrive.callFlowManager": {
+            funcName: "gpii.userListeners.tokenArrived",
+            args: [
+                "{that}",
+                "{arguments}.1" // token
+            ]
+        },
+        "onTokenRemove.callFlowManager": {
+            funcName: "gpii.userListeners.tokenRemoved",
+            args: [
+                "{that}",
+                "{arguments}.1" // token
+            ]
+        }
+    }
+});
+
+/**
+ * Handles the onTokenArrive event.
+ *
+ * It calls the "login" action for non-proximity devices, otherwise "proximityTriggered".
+ *
+ * @param that {Component} An instance of gpii.userListener.
+ * @param token {string} The token from the user listener.
+ */
+gpii.userListeners.tokenArrived = function (that, token) {
+    fluid.log(that.listenerName + " token arrived: " + token);
+
+    var action = that.proximity ? "proximityTriggered" : "login";
+    request("http://localhost:8081/user/" + token + "/" + action);
+};
+
+/**
+ * Handles the onTokenRemove event.
+ *
+ * It calls the "logout" action for non-proximity devices, otherwise it does nothing.
+ *
+ * @param that {Component} An instance of gpii.userListener.
+ * @param token {string} The token from the user listener.
+ */
+gpii.userListeners.tokenRemoved = function (that, token) {
+    fluid.log(that.listenerName + " token removed: " + token);
+
+    if (!that.proximity) {
+        request("http://localhost:8081/user/" + token + "/logout");
+    }
+};
+
+/**
+ * The listener has failed, and needs to be restarted. To avoid fast repeating failures, the restart will have a delay
+ * which increases each time.
+ *
+ * @param that {Component} An instance of gpii.userListener.
+ * @param that {Error} The error.
+ */
+gpii.userListeners.failed = function (that, err) {
+    var delay = (that.failCount++ % 10) * that.failDelay;
```

Comment:
Funky calculation - why do we set the delay back to 0 on the 10th time we fail?

Reply:
Well, I wanted it to not go above 10 - so my brain suggested the modulus op... which isn't entirely wrong. fixed it.